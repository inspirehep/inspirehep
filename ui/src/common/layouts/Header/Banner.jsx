import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Alert, Button } from 'antd';
import PropTypes from 'prop-types';

import { isNonBetaRoute } from '../../routes';
import { setBannerVisibility } from '../../../actions/ui';
import ExternalLink from '../../components/ExternalLink';

const BLOG_POST_LINK =
  'https://blog.inspirehep.net/2019/02/introducing-inspire-beta/';

class Banner extends Component {
  constructor(props) {
    super(props);
    this.onBannerAlertClose = this.onBannerAlertClose.bind(this);
  }

  onBannerAlertClose() {
    const { dispatch } = this.props;
    dispatch(setBannerVisibility(false));
  }

  render() {
    const { shouldDisplayBanner, isNonBetaPage } = this.props;
    return (
      shouldDisplayBanner &&
      !isNonBetaPage && (
        <div className="banner">
          <Alert
            type="info"
            banner
            closable
            afterClose={this.onBannerAlertClose}
            showIcon={false}
            message={
              <span>
                <strong>
                  INSPIRE beta provides a preview of new features currently
                  under development. For more info, visit our{' '}
                  <ExternalLink href={BLOG_POST_LINK}>blog</ExternalLink>. Try
                  it out and let us know what you think!
                </strong>
                <Button
                  className="ml3"
                  type="primary"
                  target="_blank"
                  href="https://goo.gl/forms/aTYSRzd7vTUhxzL43"
                >
                  Take the survey
                </Button>
              </span>
            }
          />
        </div>
      )
    );
  }
}

Banner.propTypes = {
  isNonBetaPage: PropTypes.bool.isRequired,
  shouldDisplayBanner: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
};

const stateToProps = state => ({
  isNonBetaPage: isNonBetaRoute(String(state.router.location.pathname)),
  shouldDisplayBanner: state.ui.get('bannerVisibility'),
});

const dispatchToProps = dispatch => ({ dispatch });

export default connect(stateToProps, dispatchToProps)(Banner);
