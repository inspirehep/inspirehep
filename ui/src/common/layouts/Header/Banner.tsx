import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import SanitizedHTML from 'react-sanitized-html';
import classNames from 'classnames';
import { Alert, Button } from 'antd';
import { Map } from 'immutable';

const ALLOWED_ATTRIBUTES_BY_TAG = { a: ['href', 'target', 'rel'] };
const ALLOWED_HTML_TAGS = ['a', 'p', 'em', 'strong'];

function Banner({
  type,
  closable,
  message,
  action,
  onClose,
  id,
  center,
  closedBannersById,
  currentPathname,
  pathnameRegexp,
}) {
  const afterClose = useCallback(
    () => {
      onClose(id);
    },
    [id, onClose]
  );
  const isClosed = closedBannersById.has(id);
  const shouldDisplayOnCurrentPathname = pathnameRegexp
    ? pathnameRegexp.test(currentPathname)
    : true;
  return (
    !isClosed &&
    shouldDisplayOnCurrentPathname && (
      <Alert
        type={type}
        banner
        className={classNames({ tc: center })}
        closable={closable}
        afterClose={afterClose}
        showIcon={false}
        message={
          <span>
            <SanitizedHTML
              className={classNames('di', { mr3: Boolean(action) })}
              allowedAttributes={ALLOWED_ATTRIBUTES_BY_TAG}
              allowedTags={ALLOWED_HTML_TAGS}
              html={message}
            />
            {action && (
              <Button type="primary" target="_blank" href={action.href}>
                {action.name}
              </Button>
            )}
          </span>
        }
      />
    )
  );
}

Banner.propTypes = {
  // via BANNERS config
  id: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired, // can have limited html tags
  center: PropTypes.bool,
  type: PropTypes.oneOf(['error', 'warning', 'info', 'success']),
  closable: PropTypes.bool,
  action: PropTypes.shape({
    name: PropTypes.string.isRequired,
    href: PropTypes.string.isRequired,
  }),
  pathnameRegexp: PropTypes.instanceOf(RegExp),

  // from container props
  onClose: PropTypes.func,
  closedBannersById: PropTypes.instanceOf(Map).isRequired,
  currentPathname: PropTypes.string.isRequired,
};

Banner.defaultProps = {
  type: 'info',
  closable: true,
  center: false,
};

export default Banner;
