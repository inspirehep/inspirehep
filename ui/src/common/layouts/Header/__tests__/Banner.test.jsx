import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { fromJS } from 'immutable';

import Banner from '../Banner';

describe('Banner', () => {
  it('renders when it is not closed', () => {
    const { container } = render(
      <Banner
        id="release-04.2020"
        message='<strong>Welcome to the new INSPIRE! <a href="/release">learn more</a></strong>'
        closedBannersById={fromJS({})}
        currentPathname="/"
        onClose={jest.fn()}
      />
    );

    expect(container.hasChildNodes()).toBe(true);
  });

  it('does not render when it is closed', () => {
    const { container } = render(
      <Banner
        id="release-04.2020"
        message='<strong>Welcome to the new INSPIRE! <a href="/release">learn more</a></strong>'
        closedBannersById={fromJS({ 'release-04.2020': true })}
        currentPathname="/"
        onClose={jest.fn()}
      />
    );
    expect(container.hasChildNodes()).toBe(false);
  });

  it('renders when pathname is a match', () => {
    const { container } = render(
      <Banner
        id="release-feature-04.2020"
        message="We have a new literature feature"
        pathnameRegexp={/^\/literature/}
        closedBannersById={fromJS({})}
        currentPathname="/literature/12345"
        onClose={jest.fn()}
      />
    );
    expect(container.hasChildNodes()).toBe(true);
  });

  it('does not render when pathname is not a match', () => {
    const { container } = render(
      <Banner
        id="release-feature-04.2020"
        message="We have a new feature"
        pathnameRegexp={/^\/new-feature-page/}
        closedBannersById={fromJS({})}
        currentPathname="/literature/12345"
        onClose={jest.fn()}
      />
    );
    expect(container.hasChildNodes()).toBe(false);
  });

  it('renders with custom style', () => {
    const { getByRole } = render(
      <Banner
        id="files-outage-20.04.2020"
        type="warning"
        message="We have problem with our storage."
        closable={false}
        action={{
          name: 'Learn more',
          href: 'https://status.inspirehep.net',
        }}
        center
        closedBannersById={fromJS({})}
        currentPathname="/"
        onClose={jest.fn()}
      />
    );

    const banner = getByRole('alert');
    expect(banner).toHaveClass('ant-alert-warning');
  });

  it('removes the banner after close', async () => {
    const onClose = jest.fn();
    const { container } = render(
      <Banner
        id="test"
        message="Test"
        closedBannersById={fromJS({})}
        currentPathname="/"
        onClose={onClose}
      />
    );
    const closeButton = screen.getByRole('button', { name: /close/i });
    // Simulate clicking the close button
    fireEvent.click(closeButton);
    await waitFor(() => {
      expect(container.firstChild).toBeNull();
    });
  });
});
