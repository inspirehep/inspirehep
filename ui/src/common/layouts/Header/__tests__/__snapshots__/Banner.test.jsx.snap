// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner does not render when it is closed 1`] = `""`;

exports[`Banner does not render when pathname is not a match 1`] = `""`;

exports[`Banner renders when it is not closed 1`] = `
<Alert
  afterClose={[Function]}
  banner={true}
  className=""
  closable={true}
  message={
    <span>
      <SanitizedHTML
        allowedAttributes={
          Object {
            "a": Array [
              "href",
              "target",
              "rel",
            ],
          }
        }
        allowedTags={
          Array [
            "a",
            "p",
            "em",
            "strong",
          ]
        }
        className="di"
        html="<strong>Welcome to the new INSPIRE! <a href=\\"/release\\">learn more</a></strong>"
      />
    </span>
  }
  showIcon={false}
  type="info"
/>
`;

exports[`Banner renders when pathname is a match 1`] = `
<Alert
  afterClose={[Function]}
  banner={true}
  className=""
  closable={true}
  message={
    <span>
      <SanitizedHTML
        allowedAttributes={
          Object {
            "a": Array [
              "href",
              "target",
              "rel",
            ],
          }
        }
        allowedTags={
          Array [
            "a",
            "p",
            "em",
            "strong",
          ]
        }
        className="di"
        html="We have a new literature feature"
      />
    </span>
  }
  showIcon={false}
  type="info"
/>
`;

exports[`Banner renders with custom style 1`] = `
<Alert
  afterClose={[Function]}
  banner={true}
  className="tc"
  closable={false}
  message={
    <span>
      <SanitizedHTML
        allowedAttributes={
          Object {
            "a": Array [
              "href",
              "target",
              "rel",
            ],
          }
        }
        allowedTags={
          Array [
            "a",
            "p",
            "em",
            "strong",
          ]
        }
        className="di mr3"
        html="We have problem with our storage."
      />
      <ForwardRef(InternalButton)
        href="https://status.inspirehep.net"
        target="_blank"
        type="primary"
      >
        Learn more
      </ForwardRef(InternalButton)>
    </span>
  }
  showIcon={false}
  type="warning"
/>
`;
